### PR TITLE
Adaptive threshold acceptance

### LINKED ISSUE #127: Adaptive threshold acceptance criterion
[Máximo et al. (2021)](https://arxiv.org/abs/2012.11021v1) implements a threshold-based acceptance criterion. They do not give a specific name, but the threshold changes adaptively during the search, which is why I named it "adaptive threshold". 

>This threshold is calculated according to the average quality of the solutions obtained after the local search, which is denoted by $\bar{f}$.

### PR DESCRIPTION
**Pull request by TeoSkondras**
This PR:

- [x] Closes #127.
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

### CODE REVIEW

**Comment by TeoSkondras** [comment]
@N-Wouda @leonlan  I would love to receive some feedback!

**Comment by N-Wouda** [comment]
Hi @TeoSkondras! 👋 Thanks for the PR, I'll try to have a look at this tomorrow!

**Comment by TeoSkondras** [comment]
> Hi @TeoSkondras! 👋 Thanks for the PR, I'll try to have a look at this tomorrow!

Great, I am waiting for your response. I am running the CI on my own repo because I see some style issues.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:71]: To conform to the `AcceptanceCriterion` protocol, this method should have signature:
```suggestion
    def __call__(
        self, rnd: RandomState, best: State, current: State, candidate: State
    ) -> bool:
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:48]: In English, $\eta$ is written as 'eta'. Out of curiosity: is that different in modern Greek?

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:72]: `np.append` (and further below, `np.delete`) are somewhat expensive. I think we can stick to Python lists.

Comment (N-Wouda) [reply alns/accept/AdaptiveThreshold.py:72]: Or rather, since we need to append/delete from both ends of the array, a `collections.deque`.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:79]: With a `deque`, maybe use `statistics.mean`? That's builtin and probably faster than using e.g. numpy here.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:89]:
```suggestion
        threshold = best_sol + self._eta * (avg_sol - best_sol)
        return candidate.objective() <= threshold
```

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:8]: You can get a mock like this from `alns.tests.states`, where it is called `VarObj`. There is no need to redefine one here.

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:20]: Can we also test $\eta < 0$?

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:33]: Issue: `np.arange(0, 1, -1)` is an empty array. This test does nothing.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:5]: Could you streamline the docstrings a bit more? The references and parameters sections don't really work out yet.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:60]:
```suggestion
    def eta(self) -> float:
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:64]:
```suggestion
    def gamma(self) -> float:
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:68]: I suggest we rename `solutions_received` to `history`, and type it as `List[float]` (with `typing.List`). So
```suggestion
    def history(self) -> List[float]:
```

**Review by N-Wouda — CHANGES_REQUESTED** [review]
Looks very good already, thanks! I have a few small remarks that should hopefully not take a lot of time to implement. Thanks again for the PR!

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:43]: Can we also add a test for `solutions_received`/`history`?

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:49]: Why can't $\eta$ be zero or one? The paper does not disallow that, I think. What happens when $\gamma = 0$? The paper divides by $\gamma$, so that's probably not allowed. Maybe we should require that $\gamma$ is (a) an integer, and (b) positive?

File: alns/accept/AdaptiveThreshold.py
Comment (TeoSkondras) [reply alns/accept/AdaptiveThreshold.py:48]: In English, a lot of times, an h is added in front. For example "Ερμής" is translated to "Ermes". So I thought it would be the same here hahaha.

**Comment by N-Wouda** [comment]
> @N-Wouda I just finished implementing your feedback. The only thing I did not implement is To conform to the AcceptanceCriterion protocol, this method should have the signature you showed me. I cannot understand how to call the method with this signature. If you could explain (alns.tests.states, VarObj etc) I will fix it asap.

Sure! You could have a look at the other tests (for the other acceptance criteria) to figure out how to call the method? It basically needs a few additional objects that are used by some (not all) other acceptance criteria. Here, you only need `candidate`, but the others will be passed-in by ALNS regardless.

In those other tests you should see imports like
```python
from alns.tests.states import One, Zero
```
etc. You can add `VarObj` to that list of imports: it's the same as your `MockCandidate`, and should be a drop-in replacement.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:74]:
```suggestion
    def eta(self) -> float:
        return self._eta

    @property
    def gamma(self) -> int:
        return self._gamma

    @property
    def history(self) -> List[float]:
        return self._history
```
The tests that we previously had for these properties, can we bring them back? (but then with a range that's not empty)

**Comment by TeoSkondras** [comment]
@N-Wouda Thank you for the feedback and the clarifications. I committed my last changes, let me know about anything!

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:30]:
```suggestion
    assert_equal(adaptive_threshold.eta, eta)
```

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:122]:
```suggestion
    assert_equal(adaptive_threshold.history, [7200, 7120, 7100, 7200])
```

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:14]: Can we add another test that checks $\eta > 1$? These cases only check $\eta \in \lbrace -1, 0.5 \rbrace$.

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:27]:
```suggestion
@mark.parametrize("eta", [0, 0.01, 0.5, 0.99, 1])
```

File: alns/accept/tests/test_adaptive_threshold.py
Comment (N-Wouda) [inline alns/accept/tests/test_adaptive_threshold.py:33]:
```suggestion
@mark.parametrize("gamma", range(1, 10))
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:58]:
```suggestion
        if not (0 <= eta <= 1):
            raise ValueError("eta must be in [0, 1].")
        
        if gamma <= 0:
            raise ValueError("gamma must be positive.")
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:62]: Import `collections.deque`. Then use
```suggestion
        self._history = deque(maxlen=gamma)
```

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:79]: This is no longer needed with a `deque` of `maxlen=gamma`. The deque automatically forgets old values.

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:74]: When `history` is a `deque`, I think we should return a list, like so:
```suggestion
        return list(self._history)
```

**Review by N-Wouda — COMMENTED** [review]
I think these are the last changes I have. Other than that we're nearly good to go. I will have another look at the documentation later, but that might also be after merging!

**Comment by TeoSkondras** [comment]
@N-Wouda Done I think!

File: alns/accept/AdaptiveThreshold.py
Comment (N-Wouda) [inline alns/accept/AdaptiveThreshold.py:82]:
```suggestion
```

**Review by N-Wouda — APPROVED** [review]
Can you remove the print statement? Otherwise looks good to me! Thanks for the PR! :tada:

**Comment by N-Wouda** [comment]
I'll merge this in later today. Many thanks for the PR @TeoSkondras!

**Comment by TeoSkondras** [comment]
@TeoSkondras I removed the forgotten print. My pleasure, I am really into OR (mainly VRP-type problems). I would love to exchange contact info with you for other projects!

**Comment by N-Wouda** [comment]
@TeoSkondras a bit late, my apologies, but if you like VRP-type stuff you might want to have a look at [PyVRP](https://github.com/PyVRP/PyVRP/) as well!

**Pull request merged**
